Add VendorIdVerificationSigner cert support to SDK

tcarmelveilleux opened the pull request:
#### Changes
- Add VendorIdVerificationSigner direct support to src/credentials.
  - Includes adding missing OIDs and direct methods to generate and identify such certificate types.
- Add support to the `chip-cert` tool to process/generate these certificates.
  - Includes fixing minor error-handling bugs and log wording in chip-cert.
  - Added inner CHIPLog support to avoid dropping helpful error messages on the floor.

Now it is possible to validate/generate VID Verification Signer Certificates with `chip-cert` and
directly validate them with `CHIPCert.h` API, which will be helpful for JCM (and other) client code.

#### Testing

- Added unit tests for the new methods in CHIPCert.h/.cpp module, including round-trip tests.
- Manually tested chip-cert against VVSC and RCAC in the Matter spec repos test vectors. Prior to changes, verification failed as expected.

```
tennessee@tennessee1:~/repos/connectedhomeip-spec-tcv/src/crypto_primitives/vid_verify_data$ chip-cert validate-cert --trusted-cert vid_signer_root_cert.pem vid_signer_cert.pem
tennessee@tennessee1:~/repos/connectedhomeip-spec-tcv/src/crypto_primitives/vid_verify_data$ echo $?
0

tennessee@tennessee1:~/repos/connectedhomeip-spec-tcv/src/crypto_primitives/vid_verify_data$ chip-cert convert-cert --chip vid_signer_cert.pem vid_signer_cert.tlv
tennessee@tennessee1:~/repos/connectedhomeip-spec-tcv/src/crypto_primitives/vid_verify_data$ chip-cert convert-cert --chip vid_ vid_signer_cert.tlv

tennessee@tennessee1:~/repos/connectedhomeip-spec/src/crypto_primitives/vid_verify_data$ chip-cert print-cert vid_signer_root_cert.tlv
CHIP Certificate:
    Signature Algo  : ECDSAWithSHA256
    Issuer          : [[ OrganizationName = Acme LLC,
                         CommonName = VID FFF1 Vendor verification root,
                         MatterRCACId = 11223344AABBCCDD ]]
    Not Before      : 0x2FCCC3EB  ( 2025/05/30 19:30:51 )
    Not After       : 0xEBC4E1EB  ( 2125/05/06 19:30:51 )
    Subject         : [[ OrganizationName = Acme LLC,
                         CommonName = VID FFF1 Vendor verification root,
                         MatterRCACId = 11223344AABBCCDD ]]
    Public Key Algo : ECPublicKey
    Curve Id        : prime256v1
    Public Key      : 04 C3 72 05 4D 6B 08 9B BF E4 6B CF 48 64 09 ED
                      53 73 ED C5 65 14 1C FF 5A F2 F8 32 2F 22 6D 33
                      CE 26 F4 6D E2 CA 2F 06 8B 66 46 C3 64 C2 4E 2E
                      6E 5B C1 8B 2F 2A E3 46 5A FE 32 1C D9 05 69 5D
                      53
    Extensions:
        Is CA            : true
        Path Length Const: 2
        Key Usage        : DigitalSignature KeyCertSign CRLSign
        Subject Key Id   : 10 52 A9 D8 2F 5E DB 97 A6 C4 5E CF 25 C9 91 8B A3 50 50 01
        Authority Key Id : 10 52 A9 D8 2F 5E DB 97 A6 C4 5E CF 25 C9 91 8B A3 50 50 01
    Signature       : 7F 1F 72 E6 E5 9A 7A DE 0D A9 21 DB 53 2A 7E 17
                      F0 CE 68 D3 04 46 E0 A2 8E D3 84 14 9D 46 85 1B
                      19 A5 92 50 F3 92 D8 EE 40 F0 55 BF DC 1A 87 E6
                      60 95 92 23 8D E9 AC AB E6 C7 98 F0 89 73 E4 B3
tennessee@tennessee1:~/repos/connectedhomeip-spec-tcv/src/crypto_primitives/vid_verify_data$ chip-cert print-cert  vid_signer_cert.tlv
CHIP Certificate:
    Signature Algo  : ECDSAWithSHA256
    Issuer          : [[ OrganizationName = Acme LLC,
                         CommonName = VID FFF1 Vendor verification root,
                         MatterRCACId = 11223344AABBCCDD ]]
    Not Before      : 0x2FCCC3EB  ( 2025/05/30 19:30:51 )
    Not After       : 0x4298C6EB  ( 2035/05/28 19:30:51 )
    Subject         : [[ OrganizationName = Acme LLC,
                         CommonName = VID FFF1 Vendor verification signer 1,
                         MatterVidVerificationSignerId = BBEEDDEE00112233 ]]
    Public Key Algo : ECPublicKey
    Curve Id        : prime256v1
    Public Key      : 04 9E C0 A3 6C B9 3C 01 35 79 F2 9E D3 60 4E A9
                      8E 14 40 99 36 C5 59 7A 50 14 D9 6E 1D 77 88 CB
                      55 19 F8 D8 80 6F E8 96 40 F8 A8 75 9D 4D 35 9B
                      78 30 D4 4A A2 44 E7 B8 71 DC 71 3B B5 59 12 16
                      7D
    Extensions:
        Is CA            : false
        Key Usage        : DigitalSignature
        Subject Key Id   : FA C0 E7 D0 08 94 73 F1 5C 2D FE FE 54 6F FF A0 B0 0A 8B B2
        Authority Key Id : 10 52 A9 D8 2F 5E DB 97 A6 C4 5E CF 25 C9 91 8B A3 50 50 01
    Signature       : 7E 35 E8 21 33 74 5D 43 40 99 D0 2A 3B 56 A8 7A
                      65 61 9E 63 22 FA 85 20 1A 87 72 75 5B 2A CB 16
                      AA 83 64 EE 93 63 83 40 2C 5B 91 3E 4F 40 EA C7
                      D6 92 61 66 38 B0 1B E5 DD CA 1D EF 01 3B FC 3E
```

tcarmelveilleux committed:
Add VendorIdVerificationSigner cert support to SDK

- Add VendorIdVerificationSigner direct support to src/credentials.
  - Includes adding missing OIDs and direct methods to generate
    and identify such certificate types.
- Add support to the `chip-cert` tool to process/generate these
  certificates.
  - Includes fixing minor error-handling bugs and log wording in chip-cert.
  - Added inner CHIPLog support to avoid dropping helpful error messages
    on the floor.

Testing done:

- Added unit tests for the new methods in CHIPCert.h/.cpp module,
  including round-trip tests.
- Manually tested chip-cert against VVSC and RCAC in the Matter
  spec repos test vectors. Prior to changes, verification failed
  as expected.

```
tennessee@tennessee1:~/repos/connectedhomeip-spec-tcv/src/crypto_primitives/vid_verify_data$ chip-cert validate-cert --trusted-cert vid_signer_root_cert.pem vid_signer_cert.pem
tennessee@tennessee1:~/repos/connectedhomeip-spec-tcv/src/crypto_primitives/vid_verify_data$ echo $?
0

tennessee@tennessee1:~/repos/connectedhomeip-spec-tcv/src/crypto_primitives/vid_verify_data$ chip-cert convert-cert --chip vid_signer_cert.pem vid_signer_cert.tlv
tennessee@tennessee1:~/repos/connectedhomeip-spec-tcv/src/crypto_primitives/vid_verify_data$ chip-cert convert-cert --chip vid_ vid_signer_cert.tlv

tennessee@tennessee1:~/repos/connectedhomeip-spec/src/crypto_primitives/vid_verify_data$ chip-cert print-cert vid_signer_root_cert.tlv
CHIP Certificate:
    Signature Algo  : ECDSAWithSHA256
    Issuer          : [[ OrganizationName = Acme LLC,
                         CommonName = VID FFF1 Vendor verification root,
                         MatterRCACId = 11223344AABBCCDD ]]
    Not Before      : 0x2FCCC3EB  ( 2025/05/30 19:30:51 )
    Not After       : 0xEBC4E1EB  ( 2125/05/06 19:30:51 )
    Subject         : [[ OrganizationName = Acme LLC,
                         CommonName = VID FFF1 Vendor verification root,
                         MatterRCACId = 11223344AABBCCDD ]]
    Public Key Algo : ECPublicKey
    Curve Id        : prime256v1
    Public Key      : 04 C3 72 05 4D 6B 08 9B BF E4 6B CF 48 64 09 ED
                      53 73 ED C5 65 14 1C FF 5A F2 F8 32 2F 22 6D 33
                      CE 26 F4 6D E2 CA 2F 06 8B 66 46 C3 64 C2 4E 2E
                      6E 5B C1 8B 2F 2A E3 46 5A FE 32 1C D9 05 69 5D
                      53
    Extensions:
        Is CA            : true
        Path Length Const: 2
        Key Usage        : DigitalSignature KeyCertSign CRLSign
        Subject Key Id   : 10 52 A9 D8 2F 5E DB 97 A6 C4 5E CF 25 C9 91 8B A3 50 50 01
        Authority Key Id : 10 52 A9 D8 2F 5E DB 97 A6 C4 5E CF 25 C9 91 8B A3 50 50 01
    Signature       : 7F 1F 72 E6 E5 9A 7A DE 0D A9 21 DB 53 2A 7E 17
                      F0 CE 68 D3 04 46 E0 A2 8E D3 84 14 9D 46 85 1B
                      19 A5 92 50 F3 92 D8 EE 40 F0 55 BF DC 1A 87 E6
                      60 95 92 23 8D E9 AC AB E6 C7 98 F0 89 73 E4 B3
tennessee@tennessee1:~/repos/connectedhomeip-spec-tcv/src/crypto_primitives/vid_verify_data$ chip-cert print-cert  vid_signer_cert.tlv
CHIP Certificate:
    Signature Algo  : ECDSAWithSHA256
    Issuer          : [[ OrganizationName = Acme LLC,
                         CommonName = VID FFF1 Vendor verification root,
                         MatterRCACId = 11223344AABBCCDD ]]
    Not Before      : 0x2FCCC3EB  ( 2025/05/30 19:30:51 )
    Not After       : 0x4298C6EB  ( 2035/05/28 19:30:51 )
    Subject         : [[ OrganizationName = Acme LLC,
                         CommonName = VID FFF1 Vendor verification signer 1,
                         MatterVidVerificationSignerId = BBEEDDEE00112233 ]]
    Public Key Algo : ECPublicKey
    Curve Id        : prime256v1
    Public Key      : 04 9E C0 A3 6C B9 3C 01 35 79 F2 9E D3 60 4E A9
                      8E 14 40 99 36 C5 59 7A 50 14 D9 6E 1D 77 88 CB
                      55 19 F8 D8 80 6F E8 96 40 F8 A8 75 9D 4D 35 9B
                      78 30 D4 4A A2 44 E7 B8 71 DC 71 3B B5 59 12 16
                      7D
    Extensions:
        Is CA            : false
        Key Usage        : DigitalSignature
        Subject Key Id   : FA C0 E7 D0 08 94 73 F1 5C 2D FE FE 54 6F FF A0 B0 0A 8B B2
        Authority Key Id : 10 52 A9 D8 2F 5E DB 97 A6 C4 5E CF 25 C9 91 8B A3 50 50 01
    Signature       : 7E 35 E8 21 33 74 5D 43 40 99 D0 2A 3B 56 A8 7A
                      65 61 9E 63 22 FA 85 20 1A 87 72 75 5B 2A CB 16
                      AA 83 64 EE 93 63 83 40 2C 5B 91 3E 4F 40 EA C7
                      D6 92 61 66 38 B0 1B E5 DD CA 1D EF 01 3B FC 3E
```

restyled-commits committed:
Restyled by clang-format

restyled-commits committed:
Restyled by prettier-markdown

tcarmelveilleux commented on src/tools/chip-cert/GeneralUtils.cpp:
```diff
@@ -75,7 +77,7 @@ bool InitOpenSSL()
 
     gNIDChipRootId = OBJ_create("1.3.6.1.4.1.37244.1.4", "ChipRootId", "ChipRootId");
-    if (gNIDChipICAId == 0)
+    if (gNIDChipRootId == 0)
```
Yes it is the intended fix.

andy31415 commented on src/credentials/GenerateChipX509Cert.cpp:
```diff
@@ -251,14 +258,18 @@ CHIP_ERROR EncodeExtensions(bool isCA, const Crypto::P256PublicKey & SKI, const
+            else if (certType == CertType::kVidVerificationSigner)
+            {
+                ReturnErrorOnFailure(EncodeVendorIdVerificationSignerSpecificExtensions(writer));
+            }
```
previous code used to `else` on all cases. Now we have everything fully coded in if/elseif so there is no "default" anymore if certType is some crazy type.
Would it make sense to use a `switch` for the compiler to enforce all branches taken or have an error branch with a error handle (e.g. log an error)?

tcarmelveilleux replied:
Done.

andy31415 commented on src/tools/chip-cert/Cmd_GenCert.cpp:
```diff
@@ -42,6 +42,7 @@ OptionDef gCmdOptionDefs[] =
 {
     { "type",                kArgumentRequired, 't' },
     { "subject-chip-id",     kArgumentRequired, 'i' },
+    { "subject-matter-id",   kArgumentRequired, 'i' },
```
maybe a comment saying that aliasing "i" to the same thing is applicable (that `i` reuse for the short name is not a bug).

tcarmelveilleux replied:
Done.

tcarmelveilleux committed:
Address review comments

restyled-commits committed:
Restyled by clang-format

tcarmelveilleux committed:
De-fuse a clang-tidy "code mine"

- Some bad (but legal) pattern existed in chip-cert tool source
  related to how unique_ptr<uint8_t []> are initialized.
- This blew-up with clang-tidy errors and there's more lurking.
- I fixed all of them so the next person touching chip-cert
  doesn't need to deal with it.

tcarmelveilleux committed:
Merge remote-tracking branch 'upstream/master' into add-vvs-to-chip-cert

restyled-commits committed:
Restyled by clang-format

tcarmelveilleux committed:
Merge remote-tracking branch 'upstream/master' into add-vvs-to-chip-cert

tcarmelveilleux committed:
Fix some clang-tidy left

restyled-commits committed:
Restyled by clang-format
